pkg/istio-agent/healthcheck.go

```diff
@@ -0,0 +1,227 @@
+							Code:    int32(code),
+							Message: err.Error(),
+						},
+					}
+					numSuccess = 0
+					numFail = 0
+					lastStateHealthy = false
+				}
+			}
+			// should this be a time.Ticker?
+			time.Sleep(w.config.CheckFrequency)
+		}
+	}
+
+	if w.config.CheckType == TCPHealthCheck {
+		for {
+			if err := tcpCheck(w.config.TCPConfig, w.config.ProbeTimeout); err == nil {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh && !lastStateHealthy {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+					numFail = 0
+					lastStateHealthy = true
+				}
+			} else {
+				numFail++
+				if numFail == w.config.FailThresh && lastStateHealthy {
+					notifyHealthChange <- &discovery.DiscoveryRequest{
+						TypeUrl: HealthInfoTypeURL,
+						ErrorDetail: &status.Status{
+							Code:    int32(500),
+							Message: err.Error(),
+						},
+					}
+					numSuccess = 0
+					numFail = 0
+					lastStateHealthy = false
+				}
+			}
+		}
+	}
+
+	if w.config.CheckType == ExecHealthCheck {
+		for {
+			if err := execCheck(w.config.ExecConfig); err == nil {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh && !lastStateHealthy {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+					numFail = 0
+					lastStateHealthy = false
+				} else {
+					numFail++
+					if numFail == w.config.FailThresh && lastStateHealthy {
+						notifyHealthChange <- &discovery.DiscoveryRequest{
+							TypeUrl: HealthInfoTypeURL,
+							ErrorDetail: &status.Status{
+								Code:    int32(500),
+								Message: err.Error(),
+							},
+						}
+						numSuccess = 0
+						numFail = 0
+						lastStateHealthy = false
+					}
+				}
+			}
+		}
+	}
+}
+
+// httpCheck performs a http get to a given endpoint with a timeout, and returns
+// the status and error.
+func httpCheck(config HTTPHealthCheckConfig, timeout time.Duration) (int, error) {
+	client := http.Client{
+		Timeout: timeout,
+	}
+	checkURL, err := url.Parse(fmt.Sprintf("%s://localhost:%v/%s", config.Scheme, config.Port, config.Path))
+	req := &http.Request{
+		Method: "GET",
+		URL:    checkURL,
+		Header: config.Headers,
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		resp.Body.Close()
+	}
+	return resp.StatusCode, err
+}
+
+// tcpCheck connects to a given endpoint with a timeout, and errors
+// if connecting is unavailable (unhealthy)
+func tcpCheck(config TCPHealthCheckConfig, timeout time.Duration) error {
+	d := net.Dialer{
+		Timeout: timeout,
+	}
+	url := fmt.Sprintf("%s:%v", config.Host, config.Port)
+	conn, err := d.Dial("tcp", url)
+	if conn != nil {
+		conn.Close()
+	}
+	return err
+}
+
+// todo does adding Stderr to exec cmd return the error there?
+func execCheck(config ExecHealthCheckConfig) error {
+	healthCheckCmd := &exec.Cmd{
+		Path:   config.ExecutableName,
+		Args:   config.Args,
+		Stdin:  os.Stdin,
+		Stdout: os.Stdout,
+	}
+	return healthCheckCmd.Run()
```

Comment:
we should capture the potential panic, rather than letting the wrong command crash istio agent

Reply:
you mean log the error or something? why panic?